// esbuild-plugin-glob@2.2.1/dist/index.js
'use strict';

var chokidar = require('chokidar');
var esbuild = require('esbuild');
var fastGlob = require('fast-glob');
var fs = require('fs');
var match = require('minimatch');
var path = require('path');
var invariant = require('tiny-invariant');
var unixify = require('unixify');

function _interopNamespaceDefault(e) {
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n.default = e;
  return Object.freeze(n);
}

var esbuild__namespace = /*#__PURE__*/_interopNamespaceDefault(esbuild);

function globPlugin({
  chokidarOptions,
  controls,
  silent = false,
  additionalEntrypoints = [],
  ignore = []
} = {}) {
  const log = createLogger(silent);
  const context = {
    watcher: void 0
  };
  const controlFunctions = {
    async stopWatching() {
      if (!context.watcher)
        return;
      await context.watcher.close();
    }
  };
  const plugin = {
    name: "glob",
    async setup(build) {
      if (!Array.isArray(build.initialOptions.entryPoints) || !Array.isArray(additionalEntrypoints)) {
        throw new TypeError("GlobPlugin currently only supports array entrypoints");
      }
      if (build.initialOptions.watch) {
        const entryGlobs = [...build.initialOptions.entryPoints, ...additionalEntrypoints];
        const watcher = chokidar.watch(entryGlobs, { ...chokidarOptions, ignored: ignore });
        context.watcher = watcher;
        build.initialOptions.entryPoints = void 0;
        const onRebuild = typeof build.initialOptions.watch === "object" ? build.initialOptions.watch.onRebuild : void 0;
        build.initialOptions.watch = false;
        const sharedOptions = {
          ...build.initialOptions,
          incremental: true,
          metafile: true
        };
        const entryToInputsMap = /* @__PURE__ */ new Map();
        const entryToBuildResultMap = /* @__PURE__ */ new Map();
        const entryToOutputsMap = /* @__PURE__ */ new Map();
        const matchesGlobs = (filePath) => {
          return entryGlobs.some((glob) => match(filePath, glob));
        };
        const handleBuildResult = async (entry, buildResult) => {
          invariant(watcher);
          invariant(buildResult.metafile, "Expected metafile to be created");
          const outputs = Object.keys(buildResult.metafile.outputs);
          const inputs = Object.values(buildResult.metafile.outputs).filter((output) => !!output.entryPoint).flatMap(
            (output) => Object.keys(output.inputs).filter((input) => !input.includes("node_modules")).map((input) => normalizePath(input))
          );
          watcher.add(inputs);
          entryToInputsMap.set(entry, inputs);
          entryToOutputsMap.set(entry, outputs);
          entryToBuildResultMap.set(entry, buildResult);
          onRebuild == null ? void 0 : onRebuild(null, buildResult);
        };
        const findEntriesByInput = (input) => {
          return [...entryToInputsMap.entries()].filter(([_entry, inputs]) => inputs.includes(input)).map(([entry]) => entry);
        };
        watcher.on("add", async (addedPath) => {
          if (!matchesGlobs(addedPath))
            return;
          log("[add]", addedPath);
          const buildResult = await esbuild__namespace.build({
            ...sharedOptions,
            entryPoints: [unixify(addedPath)]
          });
          handleBuildResult(addedPath, buildResult);
        }).on("change", async (changedPath) => {
          const entries = findEntriesByInput(changedPath);
          entries.forEach(async (entry) => {
            log("[change]", entry);
            const oldResult = entryToBuildResultMap.get(entry);
            try {
              invariant(oldResult == null ? void 0 : oldResult.rebuild, "Expected all build results to be incremental");
              handleBuildResult(entry, await oldResult.rebuild());
            } catch {
            }
          });
        }).on("unlink", async (unlinkedPath) => {
          if (build.initialOptions.write === false)
            return;
          const outputPaths = entryToOutputsMap.get(unlinkedPath);
          if (outputPaths) {
            log("[unlink]", unlinkedPath);
            outputPaths.forEach((outputPath) => fs.unlinkSync(outputPath));
          }
        });
      } else {
        const entryGlobs = [...build.initialOptions.entryPoints, ...additionalEntrypoints];
        const resolvedEntryPoints = await fastGlob(entryGlobs, { ignore });
        build.initialOptions.entryPoints = resolvedEntryPoints;
      }
    }
  };
  return controls ? [plugin, controlFunctions] : plugin;
}
function createLogger(silent) {
  return (...arguments_) => {
    if (!silent)
      console.log(...arguments_);
  };
}
function normalizePath(filePath) {
  return path.relative(process.cwd(), filePath.replace(/^(\w+:)/, ""));
}

exports.globPlugin = globPlugin;
